### PR TITLE
Display raw images in gallery

### PR DESCRIPTION
**Pull request by Flaburgan**
Fixes https://github.com/diaspora/diaspora/issues/7664.

As said there, I feel like when the user explicitly open an image, we should give him the original one and not a resized.

### CODE REVIEW

**Comment by denschub** [comment]
We've had too many discussion around image sizes, so I'm not even going to argue. If someone uploads an image that breaks UI/UX for some reason, we can fix that in the future. Showing the full-res image is probably fine, especially since it's clamped by the max. upload size.

However, I seriously dislike the larger thumbnails. Not only is it actually too large (the thumbnail shown is 50px wide, but the circle is 52px large), creating visible borders:

screenshot 2018-10-14 at 06 38 16

it also renderes the hover-on-thumbnail thingy kinda pointless:

screenshot 2018-10-14 at 06 40 11

The intention of those bobbles at the bottom is to give you an impression of where in the gallery one is at, and the hover is there to give you an idea on how the picture might look like. In addition, the larger thumbnails now overlap standard 3:2 photos on 16:10 and 16:9 screens, and I'd rather have small bubbles instead of obnoxious large bubbles that lay on top of photos.

**Comment by Flaburgan** [comment]
To be honest, I only discovered the `:hover` feature while doing that PR. I was using the bubbles to navigate directly to a picture, and the current size definitely is too small to be able to see what's in. I never thought of "hovering" it. That's also something that doesn't work on mobile.

For those reasons, I suggest we drop the hover feature and increase the thumb size (maybe not as big as 50px, but if you think about it it's not that big). It also needs a `background-size: cover;` to adapt the image which isn't square most of the time. The thumbnails over the picture in some cases is not a problem IMO, if the picture is very tall, a scrollbar appears.

Do you think it is a good idea?

**Comment by CSammy** [comment]
I'm always in favor of dropping features that are close to useless. It gives us more time to concentrate on important features.

**Comment by denschub** [comment]
>  I suggest we drop the hover feature

Feel free to do so.

> It also needs a background-size: cover; to adapt the image which isn't square most of the time. 

Thumbnails [will be square every time all the time](https://github.com/diaspora/diaspora/blob/21f9a9d134a7869260cb8fa08f474cf1706c6037/app/uploaders/processed_image.rb#L22-L24).

> The thumbnails over the picture in some cases is not a problem IMO

Overlapping thumbnails in the case of the most common image aspect ratio (almost every digital camera shoots pictures in either 4:3 or 3:2, which both cause overlapping) in our most common screen aspect ratios (71% of Geraspora users were using diaspora\* maximized on 16:9 and 16:10 displays a year ago) is a problem IMO.

**Comment by Flaburgan** [comment]
>Overlapping thumbnails in the case of the most common image aspect ratio (almost every digital camera shoots pictures in either 4:3 or 3:2, which both cause overlapping) in our most common screen aspect ratios (71% of Geraspora users were using diaspora* maximized on 16:9 and 16:10 displays a year ago) is a problem IMO.

Here are different mockups, to help us decide what to do.
**Current state:**
current

**With a max-height on the image:**
max-height

**Putting the thumbnails on the left:**
left
With this solution we need to be careful to not block the arrow left button when there are many thumbnails.

What do you think?

**Comment by 001101** [comment]
Would be nice if there is a option for the admin to select the quality and therefore determine if the pod has enough bandwidth for all the raw pictures to serve or if he prefers less good quality and more speed.

**Comment by SuperTux88** [comment]
The images are always loaded from the pod of the author, not the pod of the viewer. And the podmin of the author can't control the pod of the viewer. So a setting wouldn't make much sense, if a pod does have less bandwidth, it should limit the number of users (the number of possible authors).

**Comment by 001101** [comment]
So the pod of the author does control the view on the viewers pod?

**Comment by HankG** [comment]
I've played around with this versus the head of develop side by side and **really** like this.  The code changes look pretty straight forward too.  Because there is an upload size limit on images I think it looks good as is without worrying about someone having to pull down a 10MB image or something like that.

**Comment by Flaburgan** [comment]
I continue to try things on this. Especially, I'd like to correctly deals with big height images like [this one](https://user-images.githubusercontent.com/930064/47871634-e2bfe200-de0c-11e8-9bd9-8f9200babb88.png) for example, which are currently very poorly displayed on diaspora:
big-height

Unfortunately, I can't find a nice solution to display them without the thumbnails to override them: 
long-bottom

I still think about displaying them on the left (and possibly detect rtl users and display it on the right then). However, I also had an idea: those kind of images are often shared alone. Maybe simply do not display the thumbnail when there is only one image would solve 90% of use cases? @denschub would that be fine for you?

**Comment by SuperTux88** [comment]
> simply do not display the thumbnail when there is only one image

That sounds like a good idea to me anyway, because with only one image the thumbnail is pretty useless anyway.

Also I don't like the thumbnails at the side, that looks strange and probably also can collide with the arrow at the side.

**Comment by Flaburgan** [comment]
Okay PR updated, I had to do code a bit more complicated it but the result is really cool.
So first of all, no more thumbnail if there is only one image. Other improvement: the image are now display with max-height set to 90%, bottom room is for the thumbnail. So you can see the whole image but it takes as much space as there is. However, if the image is very tall (read, height is more than twice width) then the image is displayed without being resized, with a scroll bar.

screenshot_2018-11-02 diaspora
screenshot_2018-11-02 diaspora 2
screenshot_2018-11-02 diaspora 1
screenshot_2018-11-02 diaspora

I'm going to put it on https://diaspora-fr.org to see how it works and ask feedback for users.

**Comment by CSammy** [comment]
Is it possible to look at the very bottom of a very tall image without the thumbnails covering up parts of it?

File: app/assets/stylesheets/gallery.scss [hunk 1]
```diff
@@ -37,24 +48,28 @@ $thumbnail-active-size: $thumbnail-size + $thumbnail-margin;
...
-      border-radius: $thumbnail-size / 2;
 
-      &.active, &:hover{
-        margin: $thumbnail-margin / 2;
-        width: $thumbnail-active-size;
-        height: $thumbnail-active-size;
-        border-radius: $thumbnail-active-size / 2;
-        transition: linear 0.2s;
-        transition-property: height, width, margin;
+      &.active,
```
Comment (diaspora-code-review) [inline app/assets/stylesheets/gallery.scss]: Nesting should be no greater than 3, but was 4

File: app/assets/stylesheets/gallery.scss [hunk 2]
```diff
@@ -37,24 +48,28 @@ $thumbnail-active-size: $thumbnail-size + $thumbnail-margin;
...
+      &:hover {
+        opacity: 1;
+      }
+
+      &:hover::after {
```
Comment (diaspora-code-review) [inline app/assets/stylesheets/gallery.scss]: Nesting should be no greater than 3, but was 4

File: app/assets/stylesheets/gallery.scss [hunk 3]
```diff
@@ -37,24 +48,28 @@ $thumbnail-active-size: $thumbnail-size + $thumbnail-margin;
...
+        opacity: 0; // We don't want another thumbnail
+      }
+
+      &:only-child {
```
Comment (diaspora-code-review) [inline app/assets/stylesheets/gallery.scss]: Nesting should be no greater than 3, but was 4

File: app/assets/stylesheets/gallery.scss (hunk 1 above)
Comment (Flaburgan) [reply app/assets/stylesheets/gallery.scss]: I can't fix this one, when I put it outside of `.indicator` it asks me to merge it. IMO pronto should be set to 4, 3 is very low.

**Comment by Flaburgan** [comment]
>Is it possible to look at the very bottom of a very tall image without the thumbnails covering up parts of it?

Hm, yeah, I should be able to do that.

**Comment by Flaburgan** [comment]
@CSammy done

screenshot_2018-11-03 diaspora

**Review by skyschub — APPROVED** [review]
I really like this.

**Comment by denschub** [comment]
Merged!

**Pull request closed without merging**